### PR TITLE
Cache / Sparse: fix rest and enable

### PR DESCRIPTION
**Pull request by lyakh**
This should be the final (but possibly not the last) set of patches to fix all current uses of `struct comp_buffer` and to enable the incoherent cache option.

### CODE REVIEW

**Comment by lyakh** [comment]
ok, obviously this isn't working for some reason. Before the https://github.com/thesofproject/sof/pull/5976/commits/3b512ec033efab9812f532fc441e1c4c9550a847 update there were failures, that I could reproduce locally and that I fixed in that update. These new failures I cannot reproduce locally, so they will have to wait until my return from vacation in 2 weeks. But please review the bulk of this PR in the meantime - it's likely one more hidden trap that has to be fixed in the before-last commit in this PR

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@wszypelt I'm seeing the basic FW load fail again on some DUTs with empty logs. It could be the DUTs need a power cycle. Thanks.

**Comment by wszypelt** [comment]
@lgirdwood  It's true, unfortunately, the auto-login on the machine still doesn't work properly. 
The retest is on

**Comment by lgirdwood** [comment]
@lyakh can you check the CI failures on Internal CI TGL_HDA. Thanks.

**Comment by lyakh** [comment]
CI: https://sof-ci.01.org/sofpr/PR5976/build962/devicetest/ has one sof-logger check failure, and 4 suspend-resume failures on different platforms, including some with "sof-logger already dead" reason

**Comment by lyakh** [comment]
@wszypelt quickbuild failed again?

**Comment by wszypelt** [comment]
@lyakh Yes, unfortunately, the retest is already running

**Comment by lyakh** [comment]
@wszypelt some licencing / networking issues this time?
```
License checkout failed: No such feature exists.
Feature:       XTENSA_XCC_TIE
```

**Comment by wszypelt** [comment]
@lyakh Unfortunately, but after further retests, it started working properly. and unfortunately you can see problems with multicore now

**Comment by lgirdwood** [comment]
@lyakh can you drop the last patch atm, target for this is Zephyr as from previous conversations it looks like it may not be possible to force cache coherence with CC optimization of literals and other data.

**Comment by lyakh** [comment]
> @lyakh can you drop the last patch atm, target for this is Zephyr as from previous conversations it looks like it may not be possible to force cache coherence with CC optimization of literals and other data.

@lgirdwood sorry, could you explain a bit? Which previous conversation are you referring to? If forcing cache coherence is impossible, how can Zephyr be the target? Do you mean only enable it for Zephyr? Then why drop the last commit, shouldn't we modify it to only enable `CONFIG_INCOHERENT` for Zephyr? And why isn't the CC optimisation of literals and other data a problem for Zephyr - because there .data and .bss are uncached? How does that make a difference for heap data, where cached vs. uncached is managed similarly under XTOS and Zephyr?

**Comment by lyakh** [comment]
@wszypelt should we relaunch the QB test?

**Comment by lgirdwood** [comment]
@wszypelt fyi - another TGL copy failure.

File: src/audio/component.c [hunk 1]
```diff
@@ -136,8 +136,6 @@ int comp_set_state(struct comp_dev *dev, int cmd)
 
 	dev->state = requested_state;
 
-	comp_writeback(dev);
```
Comment (paulstelian97) [inline src/audio/component.c:139]: Minor typo in this commit message (coferent -> coherent)

Commit title: "cache: remove wrong component cache synchronisation"

**Review by paulstelian97 — APPROVED** [review]
I like what's going on here, just one tiny thing left to fix I believe.

File: src/ipc/ipc-helper.c [hunk 2]
```diff
@@ -292,6 +295,26 @@ int ipc_comp_free(struct ipc *ipc, uint32_t comp_id)
 		return -EINVAL;
 	}
 
+	irq_local_disable(flags);
```
Comment (lgirdwood) [inline src/ipc/ipc-helper.c:298]: native Zephyr ?

**Review by lgirdwood — COMMENTED** [review]
@lyakh what are the next steps once this is merged ? are we now ready to add SPARSE check in CI build ? @marc-hb

**Comment by marc-hb** [comment]
> are we now ready to add SPARSE check in CI build ? @marc-hb

I'm out of this, @keqiaozhang has been doing most of the work.

**Review by cujomalainey — APPROVED** [review]
Approval for google files

**Comment by lyakh** [comment]
> @lyakh what are the next steps once this is merged ? are we now ready to add SPARSE check in CI build ? @marc-hb

@lgirdwood yes, we can enable SPARSE build checks. Otherwise - just enjoy all the cached buffer advantages :-) I.e. with this buffer objects on TGL / ADL will be cached and using the coherent API

**Review by lgirdwood — APPROVED** [review]

**Comment by marc-hb** [comment]
According to @keqiaozhang and @miRoox this is causing regressio
- #6075

**Pull request merged**